target/linux/ath79/image/generic.mk

```diff
@@ -135,6 +135,21 @@ define Build/wrgg-pad-rootfs
 	$(STAGING_DIR_HOST)/bin/padjffs2 $(IMAGE_ROOTFS) -c 64 >>$@
 endef
 
+define Build/engenius_enh202-v1-factory
+	-[ -f "$@" ] && \
+	mkdir -p $@.tmp && \
+	echo '#!/bin/sh' > $@.tmp/before-upgrade.sh && \
+	echo ': > /tmp/_sys/sysupgrade.tgz' >> $@.tmp/before-upgrade.sh && \
+	$(CP) $(KDIR)/loader-enh202-v1.uImage \
+		$@.tmp/openwrt-senao-enh202-uImage-lzma.bin && \
+	$(CP) $@ $@.tmp/openwrt-senao-enh202-root.squashfs && \
+	$(TAR) -cp --numeric-owner --owner=0 --group=0 --mode=a-s --sort=name \
+		$(if $(SOURCE_DATE_EPOCH),--mtime="@$(SOURCE_DATE_EPOCH)") \
+		-C $@.tmp . > $@ && \
+	$(Build/gzip) && \
+	rm -rf $@.tmp
```

Comment:
I'm just looking at this and don't see where the additional steps compared to ar71xx/image/tiny-senao.mk and the pisen code come from. Can you please elaborate what the before-upgrade.sh is for? Or is this some leftover debugging code?

Apart from that, which of the names are determined by the OEM firmware? Just
$@.tmp/openwrt-senao-enh202-uImage-lzma.bin
$@.tmp/openwrt-senao-enh202-root.squashfs
?

(So, one could provide that one as argument and use $(DEVICE_NAME) for the other string based on the device name.)

Just answering to my questions should be enough for now ...

Reply:
>  Can you please elaborate what the before-upgrade.sh is for?

@adschm , please see [ENS202EXT - How to force factory reset during/after a factory.bin flash](https://forum.openwrt.org/t/ens202ext-how-to-force-factory-reset-during-after-a-factory-bin-flash/67522) on the forum for the details!

Reply:
(Same - but unsolved - problem exists for EMR3000 in #3152)

Reply:
> Apart from that, which of the names are determined by the OEM firmware?

Yes it's only those 2 files. ar71xx has a blank file with version as the name and I previously thought that was required but its not. OEM upgrade process only checks the uImage-lzma and root.squashfs for size and data content and then writes.

Reply:
but it also accepts the scripts before and after the upgrade
before-upgrade.sh
after-upgrade.sh